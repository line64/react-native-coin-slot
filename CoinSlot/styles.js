
export default {
  container: {
    backgroundColor: 'rgba(255,255,255,.75)',
    height: 80,
    marginHorizontal: 30,
    marginTop: 20,
    paddingVertical: 10,
    borderRadius: 60,
    flexDirection: 'row',
    justifyContent: 'space-between',
    position: 'relative'
  },
  coinImage: {
    flex: 0,
    height: 60,
    width: 60,
    marginHorizontal: 20,
    resizeMode: 'contain',
    opacity: 1
  },
  arrowImage: {
    flex: 0,
    height: 40,
    width: 70,
    resizeMode: 'contain',
    marginTop: 10,
    opacity: 1
  },
  holeImage: {
    flex: 0,
    height: 60,
    width: 60,
    marginHorizontal: 10,
    resizeMode: 'contain'
  },
  coinImageFloat: {
    flex: 1,
    height: 60,
    width: 60,
    resizeMode: 'contain',
    position: 'absolute',
    left: 0,
    top: 0,
    marginHorizontal: 20,
    marginVertical: 10,
    opacity: 0,
    transform: [{ rotation: '0deg' }]
  }
};
