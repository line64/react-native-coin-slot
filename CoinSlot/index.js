import React from 'react';
import { PanResponder, View, Image, Animated, Vibration } from 'react-native';
import { loadSound } from './utils';
import styles from './styles';

import COIN_IMAGE from './assets/coin.png';
import COIN_HOVER_IMAGE from './assets/coinHover.png';
import ARROW_IMAGE from './assets/arrow.png';
import HOLE_IMAGE from './assets/hole.png';
const COIN_SOUND = loadSound('drop.mp3');

export default class CoinSlot extends Component {

  static propTypes = {
    disabled: PropTypes.bool,
    onCoinDrop: PropTypes.func.
  };

  static defaultProps = {
    disabled: false,
    onCoinDrop: () => null,
  };

  dragCoinPosition = new Animated.ValueXY();
  dragCoinOpacity = new Animated.Value(0.1);
  dragCoinRotation = new Animated.Value(0);
  staticCoinOpacity = new Animated.Value(1);
  arrowOpacity = new Animated.Value(1);
  layoutWidth = null;
  panResponder = null;

  trimCoinPosition(dx) {
    return dx < 0 ? 0 : dx;
  }

  isAboveSlot(coinPosition) {
    return ((coinPosition+60) / layoutWidth >= .9);
  }

  insertCoinInSlot() {
    Vibration.vibrate();
    if (COIN_SOUND) COIN_SOUND.play();
    setTimeout(() => onCoinDrop(), 500);
  }

  resetCoinToStart() {
    Animated.spring(this.dragCoinPosition, { toValue: { x: 0, y: 0 } }).start();
    Animated.spring(this.arrowOpacity, { toValue: 1 }).start();
    Animated.spring(this.staticCoinOpacity, { toValue: 1 }).start();
    Animated.spring(this.dragCoinOpacity, { toValue: 0.1 }).start();
    Animated.spring(this.dragCoinRotation, { toValue: 0 }).start();
  }

  moveCoinToNewPosition(newPosition, isAboveSlot) {
    Animated.spring(this.dragCoinPosition, { toValue: { x: newPosition, y: 0 } }).start();
    Animated.spring(this.dragCoinRotation, { toValue: newPosition }).start();
    Animated.spring(this.dragCoinOpacity, { toValue: (isAboveSlot ? 0 : 1) }).start();
    Vibration.vibrate([0, 3]);
  }

  startCoinDrag() {
    Animated.spring(this.arrowOpacity, { toValue: .5 }).start();
    Animated.spring(this.staticCoinOpacity, { toValue: .25 }).start();
    Animated.spring(this.dragCoinOpacity, { toValue: 1 }).start();
  }

  viewLayoutHandler = (event) => {
    this.layoutWidth = event.nativeEvent.layout.width;
  }

  enabledCheckHandler = (event) => {
    return !this.props.disabled;
  }

  dragStartHandler = (event, gestureState) => {
    this.startCoinDrag();
  }

  dragMoveHandler = (event, gestureState) => {
    const newCoinPosition = this.trimCoinPosition(gestureState.dx);
    const isAboveSlot = this.isAboveSlot(newCoinPosition);
    this.moveCoin(newCoinPosition, isAboveSlot);
  }

  dragReleaseOrTerminateHandler = (event, gestureState) => {
    this.resetCoinToStart();
    const newCoinPosition = this.trimCoinPosition(gestureState.dx);
    const isAboveSlot = this.isAboveSlot(newCoinPosition);
    if (isAboveSlot) this.insertCoin();
  }

  componentWillMount() {
    this.panResponder = PanResponder.create({
      onStartShouldSetPanResponder: enabledCheckHandler,
      onStartShouldSetPanResponderCapture: enabledCheckHandler,
      onMoveShouldSetPanResponder: enabledCheckHandler,
      onMoveShouldSetPanResponderCapture: enabledCheckHandler,
      onPanResponderGrant: dragStartHandler,
      onPanResponderMove: dragMoveHandler,
      onPanResponderRelease: dragReleaseOrTerminateHandler,
      onPanResponderTerminate: dragReleaseOrTerminateHandler,
    });
  }

  render() {
    return (
      <View style={[ styles.container, { opacity: disable ? .5 : 1 } ]} onLayout={ this.handleViewLayout }>
        <Animated.Image
          source={ COIN_IMAGE }
          style={[ styles.coinImage, { opacity: this.staticCoinOpacity } ]}
        />
        <Animated.Image
          source={ ARROW_IMAGE }
          style={[ styles.arrowImage, { opacity: this.arrowOpacity } ]}
        />
        <Image
          source={ HOLE_IMAGE }
          style={ styles.holeImage }
        />
        <Animated.Image
          source={ COIN_HOVER_IMAGE }
          style={[
            styles.coinImageFloat,
            { opacity: this.dragCoinOpacity },
            this.dragCoinPosition.getLayout(),
            { transform: [{
              rotate: this.dragCoinRotation.interpolate({
                inputRange: [0, 120],
                outputRange: ['0deg', '360deg']
              })
            }] }
          ]}
          { ...this.panResponder.panHandlers }
        />
      </View>
    );
  }

}
