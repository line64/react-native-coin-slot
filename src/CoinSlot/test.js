import React from 'react';
import CoinSlot from './';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <CoinSlot />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
