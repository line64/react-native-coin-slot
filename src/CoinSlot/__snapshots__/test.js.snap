exports[`test renders correctly 1`] = `
<View
  onLayout={undefined}
  style={
    Array [
      Object {
        "backgroundColor": "rgba(255,255,255,.75)",
        "borderRadius": 60,
        "flexDirection": "row",
        "height": 80,
        "justifyContent": "space-between",
        "marginHorizontal": 30,
        "marginTop": 20,
        "paddingVertical": 10,
        "position": "relative",
      },
      Object {
        "opacity": 1,
      },
    ]
  }>
  <Image
    source={1}
    style={
      Object {
        "flex": 0,
        "height": 60,
        "marginHorizontal": 20,
        "opacity": 1,
        "resizeMode": "contain",
        "width": 60,
      }
    } />
  <Image
    source={1}
    style={
      Object {
        "flex": 0,
        "height": 40,
        "marginTop": 10,
        "opacity": 1,
        "resizeMode": "contain",
        "width": 70,
      }
    } />
  <Image
    source={1}
    style={
      Object {
        "flex": 0,
        "height": 60,
        "marginHorizontal": 10,
        "resizeMode": "contain",
        "width": 60,
      }
    } />
  <Image
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    source={1}
    style={
      Object {
        "flex": 1,
        "height": 60,
        "left": 0,
        "marginHorizontal": 20,
        "marginVertical": 10,
        "opacity": 0.1,
        "position": "absolute",
        "resizeMode": "contain",
        "top": 0,
        "transform": Array [
          Object {
            "rotate": "0deg",
          },
        ],
        "width": 60,
      }
    } />
</View>
`;
